src/qtism/common/datatypes/QtiCoords.php

```diff
@@ -120,52 +120,58 @@ public function inside(QtiPoint $point)
     {
         if ($this->getShape() === QtiShape::DEF) {
             return true;
-        } elseif ($this->getShape() === QtiShape::RECT) {
+        }
+
+        if ($this->getShape() === QtiShape::RECT) {
             return $point->getX() >= $this[0] && $point->getX() <= $this[2] && $point->getY() >= $this[1] && $point->getY() <= $this[3];
-        } elseif ($this->getShape() === QtiShape::CIRCLE) {
+        }
+
+        if ($this->getShape() === QtiShape::CIRCLE) {
             return pow($point->getX() - $this[0], 2) + pow($point->getY() - $this[1], 2) < pow($this[2], 2);
-        } else {
-            // we consider it is a polygon.
-            // - Transform coordinates in vertices.
-            // -- Use of the "point in polygon" algorithm.
-            $vertices = [];
-            for ($i = 0; $i < count($this); $i++) {
-                $vertex = [];
-                $vertex[] = $this[$i]; //x
-                $i++;
-                $vertex[] = $this[$i]; //y
-
-                $vertices[] = $vertex;
+        }
+
+        // we consider it is a polygon.
+        // - Transform coordinates in vertices.
+        // -- Use of the "point in polygon" algorithm.
+        $vertices = [];
+        for ($i = 0, $limit = $this->count(); $i < $limit; $i++) {
+            $vertex = [];
+            $vertex[] = $this[$i]; //x
+            $i++;
+            $vertex[] = $this[$i]; //y
+
+            $vertices[] = $vertex;
+        }
+
+        $intersects = 0;
+        for ($limit = count($vertices), $i = 0, $j = $limit - 1; $i < $limit; $i++) {
+            $vertex1 = $vertices[$i];
+            $vertex2 = $vertices[$j];
+
+            if ($vertex1[1] === $vertex2[1] && $vertex1[1] === $point->getY() && $point->getX() > min($vertex1[0], $vertex2[0]) && $point->getX() < max($vertex1[0], $vertex2[0])) {
+                // we are on a boundary.
+                return true;
             }
 
-            $intersects = 0;
-            for ($i = 1; $i < count($vertices); $i++) {
```

Comment:
The problem was here.
It didn't check the edge between $vertices[count($vertices) - 1] and `$vertices[0]`, that's it.